DXMainClient/DXGUI/Multiplayer/GameLobby/MapCodeHelper.cs

```diff
@@ -20,9 +22,102 @@ public static void ApplyMapCode(IniFile mapIni, string customIniPath, GameMode g
             if (gameMode != null)
                 extraIniName = associatedIni.GetStringValue("GameModeIncludes", gameMode.Name, null);
             associatedIni.EraseSectionKeys("GameModeIncludes");
-            IniFile.ConsolidateIniFiles(mapIni, associatedIni);
+            ApplyMapCode(mapIni, associatedIni);
             if (!String.IsNullOrEmpty(extraIniName))
-                IniFile.ConsolidateIniFiles(mapIni, new IniFile(ProgramConstants.GamePath + extraIniName));
+                ApplyMapCode(mapIni, new IniFile(ProgramConstants.GamePath + extraIniName));
+        }
+
+        /// <summary>
+        /// Apply map code from an arbitrary INI file to a map INI file.
+        /// </summary>
+        /// <param name="mapIni">The map INI file.</param>
+        /// <param name="mapCodeIni">The INI file to apply to map INI file.</param>
+        public static void ApplyMapCode(IniFile mapIni, IniFile mapCodeIni)
+        {
+            ReplaceMapObjects(mapIni, mapCodeIni, "Aircraft");
+            ReplaceMapObjects(mapIni, mapCodeIni, "Infantry");
+            ReplaceMapObjects(mapIni, mapCodeIni, "Units");
+            ReplaceMapObjects(mapIni, mapCodeIni, "Structures");
+            ReplaceMapObjects(mapIni, mapCodeIni, "Terrain");
+            IniFile.ConsolidateIniFiles(mapIni, mapCodeIni);
+        }
+
+        /// <summary>
+        /// Replace all instances of objects defined in specific map section that match ID's with new object ID's.
+        /// </summary>
+        /// <param name="mapIni">The map INI file.</param>
+        /// <param name="mapCodeIni">The INI file to apply to map INI file.</param>
+        /// <param name="sectionName">The object section ID.</param>
+        private static void ReplaceMapObjects(IniFile mapIni, IniFile mapCodeIni, string sectionName)
+        {
+            string replaceSectionName = "ReplaceMap" + sectionName;
+
+            List<KeyValuePair<string, string>> objectRemapPairs = GetKeyValuePairs(mapIni, replaceSectionName);
```

Comment:
That is an actual oversight and error. It does not stop it from working because client comes back to the code for second time when it tries to apply GlobalCode.ini on the map and by then contents of the game mode's INI have already been merged into the map.